### ISSUE TITLE
Allow the execution and debug of sub-tests via VSCode UI

### ISSUE THREAD

**Issue by devuo** [golang/vscode-go#2536]
**Is your feature request related to a problem? Please describe.**

See below for concrete details of the problem and request.

**Describe the solution you'd like**
The solution I'd like is for vscode-go to offer the same kind of UX that it currently provides at root test level, but also for sub-tests like JetBrains GoLand, that other languages and test runners in VSCode have.

Here's what it looks like in GoLand:

<img width="653" alt="Screenshot 2022-11-17 at 16 36 26" src="https://user-images.githubusercontent.com/101122/202504304-868f1de0-345b-4e61-9cba-9a105aa2f385.png">
<img width="654" alt="JetBrains GoLand" src="https://user-images.githubusercontent.com/101122/202503632-4964527d-82a9-4f73-81a2-03dacb5e6ed7.png">

And here's how the same file looks like in VSCode:

<img width="636" alt="Screenshot 2022-11-17 at 16 38 18" src="https://user-images.githubusercontent.com/101122/202504608-f7d2d9f2-95be-4717-adc4-839f04bb33d2.png">

There's no way via the UI to run a sub-test other than with "Go: Subtest At Cursor" command and no way whatsoever to run it in debug mode.

The request therefore is to add support in vscode-go for the same UI buttons that appear in root tests (specifically, "run test" and "debug test") to also appear for sub tests, like in GoLand.

Here's an example of the same exact thing in VSCode but provided by [Jest Runner](https://marketplace.visualstudio.com/items?itemName=firsttris.vscode-jest-runner) for Jest tests: 

<img width="371" alt="VSCode Jest Runner" src="https://user-images.githubusercontent.com/101122/202503491-579100b6-f4ae-4a2f-a84e-c365ff460d66.png">

**Comment by devuo** [comment]
Created a PR with a working solution for this issue: https://github.com/golang/vscode-go/pull/2537

<img width="586" alt="Screenshot 2022-11-17 at 19 30 03" src="https://user-images.githubusercontent.com/101122/202540900-cf448cb7-041e-446f-9076-8fde521af6f2.png">

**Comment by hyangah** [comment]
Test explorer (the beaker icon) provides access to the subtests and that works not only with the subtests with simple string names, but subtests with generated test names.

**Comment by devuo** [comment]
The limitation of the test explorer is that it's only aware of the tests only after you run them:

Before the main test is executed:
<img width="443" alt="Screenshot 2022-11-18 at 10 33 04" src="https://user-images.githubusercontent.com/101122/202682790-a75b90b3-50a6-47ad-b137-6f173d5d8ae5.png">

After the test is executed:
<img width="441" alt="Screenshot 2022-11-18 at 10 33 26" src="https://user-images.githubusercontent.com/101122/202682856-f5fe9aab-2449-4fc3-9c4c-60a1392cdd0a.png">

My guess without even looking at code from the behavior I can see, is that it parses the tests results output here to generate the test tree we see there:
<img width="1019" alt="Screenshot 2022-11-18 at 10 35 03" src="https://user-images.githubusercontent.com/101122/202683305-9a376ed1-c26f-4c55-9b80-4da0a9a0119c.png">

Because as soon as I make *any* change to the test file, the sub tests disappear from the test explorer and need to re-run the entire root test for them to appear, which defeats the whole purpose.

**Comment by devuo** [comment]
The solution I presented, while admittedly not bulletproof in anyway, follows exactly the same logic that the extension's is using for the `subtest at cursor` command: https://github.com/golang/vscode-go/blob/1da1ea2c9775549dbe21cbba17845993d5fddaa9/src/goTest.ts#L182-L195

**Comment by devuo** [comment]
Meanwhile to solve the brittleness issues with regular expressions to find the tests and sub tests I implemented a small CLI tool that parses the AST of a go test source file and returns a JSON outline of all tests and subtests within: https://github.com/devuo/gotestoutline

@hyangah let me know if you would be happy with this approach instead of the regular expressions and I'll implement re-open a PR with something based off the `gotestoutline` instead.

**Comment by hyangah** [comment]
Thanks @devuo Eventually, the codelens logic belongs to gopls (see the [existing test codelens code in gopls]( https://github.com/golang/tools/blob/b978661c6cbaacda12e290b1e88c7675e49452dd/gopls/internal/lsp/source/code_lens.go#L38). The current version of extension doesn't use this logic yet, but we will eventually let gopls produces test code lenses. (You can test it by setting "gopls.ui.codelenses" "test" codelens).

Meanwhile, to get the PR 2537 going

* Can you follow the instruction the gopherbot added in the PR description? 
* Can you make the codelens simply call the command corresponding to "Go: Subtest At Cursor"? 

Still I see incorrect subtest arg generation - in case of nested subtests. But that is a bug in "Go: Subtest At Cursor", which can be addressed in a separate issue or as we migrate to `gopls` based codelens.

**Comment by devuo** [comment]
@hyangah made the requested changes (and a few more as I found some issues while manually testing). I'll wait for your input and OK in the implemented approach before investing time in writing/fixing new/existing tests.
